File: sdks/python/fluxum/_msgpack.py
"""A minimal MessagePack codec — just the subset the Fluxum wire uses.

The SDK ships zero runtime dependencies (SDK-080), so rather than pull in an
external MessagePack package it carries the small codec it needs: the
`[tag, payload]` envelopes (RPC-011) and the values inside them (nil, bool,
ints of every width, float, str, bin, array, map). Framing and FluxBIN rows
are handled elsewhere (`protocol`, `fluxbin`); this is only the envelope
serialization Thunder leaves to the product.

Encoding is deliberately canonical-enough for the wire, not minimal-width
byte-golf: the server decodes positionally and does not care whether an id
came as `uint8` or `uint32`, so we pick a correct width, not the smallest.
"""

from __future__ import annotations

import struct
from typing import Any, Tuple


class MsgpackError(ValueError):
    """Raised when bytes are not valid MessagePack (or exceed our subset)."""


# --- encode ------------------------------------------------------------------


def encode(value: Any) -> bytes:
    """Encode `value` to MessagePack bytes."""
    out = bytearray()
    _encode_into(value, out)
    return bytes(out)


def _encode_into(value: Any, out: bytearray) -> None:
    # bool BEFORE int: bool is a subclass of int in Python and would otherwise
    # serialize as 0/1 integers.
    if value is None:
        out.append(0xC0)
    elif value is True:
        out.append(0xC3)
    elif value is False:
        out.append(0xC2)
    elif isinstance(value, int):
        _encode_int(value, out)
    elif isinstance(value, float):
        out.append(0xCB)
        out.extend(struct.pack(">d", value))
    elif isinstance(value, str):
        _encode_str(value, out)
    elif isinstance(value, (bytes, bytearray)):
        _encode_bin(bytes(value), out)
    elif isinstance(value, (list, tuple)):
        _encode_array(value, out)
    elif isinstance(value, dict):
        _encode_map(value, out)
    else:
        raise MsgpackError(f"cannot encode {type(value).__name__}")


def _encode_int(n: int, out: bytearray) -> None:
    if 0 <= n <= 0x7F:
        out.append(n)  # positive fixint
    elif -32 <= n < 0:
        out.append(n & 0xFF)  # negative fixint
    elif n >= 0:
        if n <= 0xFF:
            out.append(0xCC)
            out.append(n)
        elif n <= 0xFFFF:
            out.append(0xCD)
            out.extend(struct.pack(">H", n))
        elif n <= 0xFFFF_FFFF:
            out.append(0xCE)
            out.extend(struct.pack(">I", n))
        elif n <= 0xFFFF_FFFF_FFFF_FFFF:
            out.append(0xCF)
            out.extend(struct.pack(">Q", n))
        else:
            raise MsgpackError("integer above u64::MAX")
    else:
        if n >= -(1 << 7):
            out.append(0xD0)
            out.extend(struct.pack(">b", n))
        elif n >= -(1 << 15):
            out.append(0xD1)
            out.extend(struct.pack(">h", n))
        elif n >= -(1 << 31):
            out.append(0xD2)
            out.extend(struct.pack(">i", n))
        elif n >= -(1 << 63):
            out.append(0xD3)
            out.extend(struct.pack(">q", n))
        else:
            raise MsgpackError("integer below i64::MIN")


def _encode_str(s: str, out: bytearray) -> None:
    data = s.encode("utf-8")
    n = len(data)
    if n <= 31:
        out.append(0xA0 | n)  # fixstr
    elif n <= 0xFF:
        out.append(0xD9)
        out.append(n)
    elif n <= 0xFFFF:
        out.append(0xDA)
        out.extend(struct.pack(">H", n))
    else:
        out.append(0xDB)
        out.extend(struct.pack(">I", n))
    out.extend(data)


def _encode_bin(data: bytes, out: bytearray) -> None:
    n = len(data)
    if n <= 0xFF:
        out.append(0xC4)
        out.append(n)
    elif n <= 0xFFFF:
        out.append(0xC5)
        out.extend(struct.pack(">H", n))
    else:
        out.append(0xC6)
        out.extend(struct.pack(">I", n))
    out.extend(data)


def _encode_array(items: Any, out: bytearray) -> None:
    n = len(items)
    if n <= 15:
        out.append(0x90 | n)
    elif n <= 0xFFFF:
        out.append(0xDC)
        out.extend(struct.pack(">H", n))
    else:
        out.append(0xDD)
        out.extend(struct.pack(">I", n))
    for item in items:
        _encode_into(item, out)


def _encode_map(mapping: dict, out: bytearray) -> None:
    n = len(mapping)
    if n <= 15:
        out.append(0x80 | n)
    elif n <= 0xFFFF:
        out.append(0xDE)
        out.extend(struct.pack(">H", n))
    else:
        out.append(0xDF)
        out.extend(struct.pack(">I", n))
    for key, val in mapping.items():
        _encode_into(key, out)
        _encode_into(val, out)


# --- decode ------------------------------------------------------------------


def decode(data: bytes) -> Any:
    """Decode a single MessagePack value; the whole buffer must be consumed."""
    value, offset = _decode_at(data, 0)
    if offset != len(data):
        raise MsgpackError(f"{len(data) - offset} trailing byte(s) after value")
    return value


def _decode_at(data: bytes, offset: int) -> Tuple[Any, int]:
    if offset >= len(data):
        raise MsgpackError("unexpected end of buffer")
    b = data[offset]
    offset += 1

    if b <= 0x7F:
        return b, offset  # positive fixint
    if b >= 0xE0:
        return b - 0x100, offset  # negative fixint
    if 0x80 <= b <= 0x8F:
        return _decode_map(data, offset, b & 0x0F)
    if 0x90 <= b <= 0x9F:
        return _decode_array(data, offset, b & 0x0F)
    if 0xA0 <= b <= 0xBF:
        return _decode_str(data, offset, b & 0x1F)

    if b == 0xC0:
        return None, offset
    if b == 0xC2:
        return False, offset
    if b == 0xC3:
        return True, offset
    if b == 0xC4:
        n = data[offset]
        return _take(data, offset + 1, n)
    if b == 0xC5:
        (n,) = struct.unpack_from(">H", data, offset)
        return _take(data, offset + 2, n)
    if b == 0xC6:
        (n,) = struct.unpack_from(">I", data, offset)
        return _take(data, offset + 4, n)
    if b == 0xCA:
        (v,) = struct.unpack_from(">f", data, offset)
        return v, offset + 4
    if b == 0xCB:
        (v,) = struct.unpack_from(">d", data, offset)
        return v, offset + 8
    if b == 0xCC:
        return data[offset], offset + 1
    if b == 0xCD:
        (v,) = struct.unpack_from(">H", data, offset)
        return v, offset + 2
    if b == 0xCE:
        (v,) = struct.unpack_from(">I", data, offset)
        return v, offset + 4
    if b == 0xCF:
        (v,) = struct.unpack_from(">Q", data, offset)
        return v, offset + 8
    if b == 0xD0:
        (v,) = struct.unpack_from(">b", data, offset)
        return v, offset + 1
    if b == 0xD1:
        (v,) = struct.unpack_from(">h", data, offset)
        return v, offset + 2
    if b == 0xD2:
        (v,) = struct.unpack_from(">i", data, offset)
        return v, offset + 4
    if b == 0xD3:
        (v,) = struct.unpack_from(">q", data, offset)
        return v, offset + 8
    if b == 0xD9:
        n = data[offset]
        return _decode_str(data, offset + 1, n)
    if b == 0xDA:
        (n,) = struct.unpack_from(">H", data, offset)
        return _decode_str(data, offset + 2, n)
    if b == 0xDB:
        (n,) = struct.unpack_from(">I", data, offset)
        return _decode_str(data, offset + 4, n)
    if b == 0xDC:
        (n,) = struct.unpack_from(">H", data, offset)
        return _decode_array(data, offset + 2, n)
    if b == 0xDD:
        (n,) = struct.unpack_from(">I", data, offset)
        return _decode_array(data, offset + 4, n)
    if b == 0xDE:
        (n,) = struct.unpack_from(">H", data, offset)
        return _decode_map(data, offset + 2, n)
    if b == 0xDF:
        (n,) = struct.unpack_from(">I", data, offset)
        return _decode_map(data, offset + 4, n)

    raise MsgpackError(f"unsupported MessagePack byte 0x{b:02x}")


def _take(data: bytes, offset: int, n: int) -> Tuple[bytes, int]:
    end = offset + n
    if end > len(data):
        raise MsgpackError("bin/str length exceeds buffer")
    return data[offset:end], end


def _decode_str(data: bytes, offset: int, n: int) -> Tuple[str, int]:
    raw, end = _take(data, offset, n)
    try:
        return raw.decode("utf-8"), end
    except UnicodeDecodeError as exc:
        raise MsgpackError("string is not valid UTF-8") from exc


def _decode_array(data: bytes, offset: int, n: int) -> Tuple[list, int]:
    items = []
    for _ in range(n):
        item, offset = _decode_at(data, offset)
        items.append(item)
    return items, offset


def _decode_map(data: bytes, offset: int, n: int) -> Tuple[dict, int]:
    out = {}
    for _ in range(n):
        key, offset = _decode_at(data, offset)
        val, offset = _decode_at(data, offset)
        out[key] = val
    return out, offset

File: sdks/csharp/Fluxum.Sdk/MsgPack.cs
// A minimal MessagePack codec — just the subset the Fluxum wire uses.
//
// The SDK ships no third-party dependencies (SDK-080), so rather than pull in
// an external MessagePack package it carries the small codec it needs: the
// [tag, payload] envelopes (RPC-011) and the values inside them (nil, bool,
// ints of every width, float, str, bin, array, map). Framing and FluxBIN rows
// live elsewhere (Protocol.cs, FluxBin.cs).

using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.Text;

namespace Fluxum.Sdk;

internal static class MsgPack
{
    public static byte[] Encode(object? value)
    {
        var buf = new List<byte>();
        EncodeInto(value, buf);
        return buf.ToArray();
    }

    private static void EncodeInto(object? value, List<byte> outb)
    {
        switch (value)
        {
            case null:
                outb.Add(0xC0);
                break;
            case bool b:
                outb.Add(b ? (byte)0xC3 : (byte)0xC2);
                break;
            case byte[] bin:
                EncodeBin(bin, outb);
                break;
            case string s:
                EncodeStr(s, outb);
                break;
            case double d:
                outb.Add(0xCB);
                AppendU64(outb, BitConverter.DoubleToUInt64Bits(d));
                break;
            case float f:
                outb.Add(0xCB);
                AppendU64(outb, BitConverter.DoubleToUInt64Bits(f));
                break;
            case ulong ul:
                EncodeUint(ul, outb);
                break;
            case long l:
                EncodeInt(l, outb);
                break;
            case int i:
                EncodeInt(i, outb);
                break;
            case uint ui:
                EncodeUint(ui, outb);
                break;
            case System.Collections.IEnumerable list when value is not string:
                var items = new List<object?>();
                foreach (var it in list) items.Add(it);
                EncodeArray(items, outb);
                break;
            default:
                throw new FormatException($"msgpack: cannot encode {value.GetType().Name}");
        }
    }

    private static void EncodeInt(long n, List<byte> outb)
    {
        if (n >= 0) { EncodeUint((ulong)n, outb); return; }
        if (n >= -32) { outb.Add((byte)n); return; }
        if (n >= sbyte.MinValue) { outb.Add(0xD0); outb.Add((byte)(sbyte)n); return; }
        if (n >= short.MinValue) { outb.Add(0xD1); AppendU16(outb, (ushort)(short)n); return; }
        if (n >= int.MinValue) { outb.Add(0xD2); AppendU32(outb, (uint)(int)n); return; }
        outb.Add(0xD3); AppendU64(outb, (ulong)n);
    }

    private static void EncodeUint(ulong n, List<byte> outb)
    {
        if (n <= 0x7F) { outb.Add((byte)n); return; }
        if (n <= 0xFF) { outb.Add(0xCC); outb.Add((byte)n); return; }
        if (n <= 0xFFFF) { outb.Add(0xCD); AppendU16(outb, (ushort)n); return; }
        if (n <= 0xFFFFFFFF) { outb.Add(0xCE); AppendU32(outb, (uint)n); return; }
        outb.Add(0xCF); AppendU64(outb, n);
    }

    private static void EncodeStr(string s, List<byte> outb)
    {
        var data = Encoding.UTF8.GetBytes(s);
        int n = data.Length;
        if (n <= 31) outb.Add((byte)(0xA0 | n));
        else if (n <= 0xFF) { outb.Add(0xD9); outb.Add((byte)n); }
        else if (n <= 0xFFFF) { outb.Add(0xDA); AppendU16(outb, (ushort)n); }
        else { outb.Add(0xDB); AppendU32(outb, (uint)n); }
        outb.AddRange(data);
    }

    private static void EncodeBin(byte[] data, List<byte> outb)
    {
        int n = data.Length;
        if (n <= 0xFF) { outb.Add(0xC4); outb.Add((byte)n); }
        else if (n <= 0xFFFF) { outb.Add(0xC5); AppendU16(outb, (ushort)n); }
        else { outb.Add(0xC6); AppendU32(outb, (uint)n); }
        outb.AddRange(data);
    }

    private static void EncodeArray(List<object?> items, List<byte> outb)
    {
        int n = items.Count;
        if (n <= 15) outb.Add((byte)(0x90 | n));
        else if (n <= 0xFFFF) { outb.Add(0xDC); AppendU16(outb, (ushort)n); }
        else { outb.Add(0xDD); AppendU32(outb, (uint)n); }
        foreach (var it in items) EncodeInto(it, outb);
    }

    private static void AppendU16(List<byte> outb, ushort v)
    {
        Span<byte> t = stackalloc byte[2];
        BinaryPrimitives.WriteUInt16BigEndian(t, v);
        outb.Add(t[0]); outb.Add(t[1]);
    }

    private static void AppendU32(List<byte> outb, uint v)
    {
        Span<byte> t = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(t, v);
        for (int i = 0; i < 4; i++) outb.Add(t[i]);
    }

    private static void AppendU64(List<byte> outb, ulong v)
    {
        Span<byte> t = stackalloc byte[8];
        BinaryPrimitives.WriteUInt64BigEndian(t, v);
        for (int i = 0; i < 8; i++) outb.Add(t[i]);
    }

    // Decode a single value; the whole buffer must be consumed. Numbers surface
    // as long (or ulong above the long range); binary as byte[]; strings as
    // string; arrays as List<object?>; maps as Dictionary<string, object?>.
    public static object? Decode(byte[] data)
    {
        var (value, off) = DecodeAt(data, 0);
        if (off != data.Length)
            throw new FormatException($"msgpack: {data.Length - off} trailing byte(s)");
        return value;
    }

    private static (object?, int) DecodeAt(byte[] data, int off)
    {
        if (off >= data.Length) throw new FormatException("msgpack: unexpected end of buffer");
        byte b = data[off++];

        if (b <= 0x7F) return ((long)b, off);
        if (b >= 0xE0) return ((long)(sbyte)b, off);
        if (b >= 0x80 && b <= 0x8F) return DecodeMap(data, off, b & 0x0F);
        if (b >= 0x90 && b <= 0x9F) return DecodeArray(data, off, b & 0x0F);
        if (b >= 0xA0 && b <= 0xBF) return DecodeStr(data, off, b & 0x1F);

        switch (b)
        {
            case 0xC0: return (null, off);
            case 0xC2: return (false, off);
            case 0xC3: return (true, off);
            case 0xC4: return DecodeBin(data, off + 1, data[off]);
            case 0xC5: return DecodeBin(data, off + 2, BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(off)));
            case 0xC6: return DecodeBin(data, off + 4, (int)BinaryPrimitives.ReadUInt32BigEndian(data.AsSpan(off)));
            case 0xCA: return ((double)BitConverter.UInt32BitsToSingle(BinaryPrimitives.ReadUInt32BigEndian(data.AsSpan(off))), off + 4);
            case 0xCB: return (BitConverter.UInt64BitsToDouble(BinaryPrimitives.ReadUInt64BigEndian(data.AsSpan(off))), off + 8);
            case 0xCC: return ((long)data[off], off + 1);
            case 0xCD: return ((long)BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(off)), off + 2);
            case 0xCE: return ((long)BinaryPrimitives.ReadUInt32BigEndian(data.AsSpan(off)), off + 4);
            case 0xCF:
                ulong u = BinaryPrimitives.ReadUInt64BigEndian(data.AsSpan(off));
                return (u <= long.MaxValue ? (long)u : u, off + 8);
            case 0xD0: return ((long)(sbyte)data[off], off + 1);
            case 0xD1: return ((long)BinaryPrimitives.ReadInt16BigEndian(data.AsSpan(off)), off + 2);
            case 0xD2: return ((long)BinaryPrimitives.ReadInt32BigEndian(data.AsSpan(off)), off + 4);
            case 0xD3: return (BinaryPrimitives.ReadInt64BigEndian(data.AsSpan(off)), off + 8);
            case 0xD9: return DecodeStr(data, off + 1, data[off]);
            case 0xDA: return DecodeStr(data, off + 2, BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(off)));
            case 0xDB: return DecodeStr(data, off + 4, (int)BinaryPrimitives.ReadUInt32BigEndian(data.AsSpan(off)));
            case 0xDC: return DecodeArray(data, off + 2, BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(off)));
            case 0xDD: return DecodeArray(data, off + 4, (int)BinaryPrimitives.ReadUInt32BigEndian(data.AsSpan(off)));
            case 0xDE: return DecodeMap(data, off + 2, BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(off)));
            case 0xDF: return DecodeMap(data, off + 4, (int)BinaryPrimitives.ReadUInt32BigEndian(data.AsSpan(off)));
        }
        throw new FormatException($"msgpack: unsupported byte 0x{b:x2}");
    }

    private static (object?, int) DecodeBin(byte[] data, int off, int n)
    {
        int end = off + n;
        if (end > data.Length) throw new FormatException("msgpack: bin length exceeds buffer");
        var outb = new byte[n];
        Array.Copy(data, off, outb, 0, n);
        return (outb, end);
    }

    private static (object?, int) DecodeStr(byte[] data, int off, int n)
    {
        int end = off + n;
        if (end > data.Length) throw new FormatException("msgpack: str length exceeds buffer");
        return (Encoding.UTF8.GetString(data, off, n), end);
    }

    private static (object?, int) DecodeArray(byte[] data, int off, int n)
    {
        var items = new List<object?>(n);
        for (int i = 0; i < n; i++)
        {
            var (item, next) = DecodeAt(data, off);
            items.Add(item);
            off = next;
        }
        return (items, off);
    }

    private static (object?, int) DecodeMap(byte[] data, int off, int n)
    {
        var outb = new Dictionary<string, object?>(n);
        for (int i = 0; i < n; i++)
        {
            var (key, koff) = DecodeAt(data, off);
            var (val, voff) = DecodeAt(data, koff);
            if (key is string ks) outb[ks] = val;
            off = voff;
        }
        return (outb, off);
    }
}
